### PR TITLE
Withdraw liquidity

### PR DESCRIPTION
**Pull request by sipec**

### CODE REVIEW

File: backend/api/src/on-update-liquidity-provision.ts [hunk 1]
```diff
@@ -24,6 +24,8 @@ export const onCreateLiquidityProvision = async (
 
   log(`onCreateLiquidityProvision: ${JSON.stringify(liquidity)}`)
 
+  if (liquidity.amount <= 0) return
```
Comment (mantikoros) [inline backend/api/src/on-update-liquidity-provision.ts:27]: Change the min amount to 1?

File: backend/api/src/remove-liquidity.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    })
+
+    await runTxn(pgTrans, {
+      fromType: 'CONTRACT',
+      fromId: contract.id,
+      toType: 'USER',
+      toId: auth.uid,
+      category: 'REMOVE_SUBSIDY',
+      amount: totalAmount,
+      token: contract.token === 'CASH' ? 'CASH' : 'M$',
```
Comment (mantikoros) [inline backend/api/src/remove-liquidity.ts:83]: why not just contract.token

File: common/src/calculate-cpmm.ts [hunk 3]
```diff
@@ -577,14 +576,67 @@ export function addCpmmMultiLiquidityAnswersSumToOne(
...
+  pool: { [outcome: string]: number },
+  p: number,
+  amount: number
+) {
+  const { newPool, liquidity, newP } = addCpmmLiquidity(pool, p, -1 * amount)
+
+  let error = ''
+
+  if (newPool.YES < MINIMUM_LIQUIDITY || newPool.NO < MINIMUM_LIQUIDITY) {
+    error = 'below minimum'
```
Comment (mantikoros) [inline common/src/calculate-cpmm.ts:594]: we don't really use this style of error handling (versus throwing an exception) anywhere else on the backend to my knowledge...

File: common/src/calculate-cpmm.ts [hunk 4]
```diff
@@ -577,14 +576,67 @@ export function addCpmmMultiLiquidityAnswersSumToOne(
...
+}
+
+// https://www.wolframalpha.com/input?i=z+%3C+%28y+-+b%29+%2F+%28%28y-b%29+%2B+%28n-b%29%29+%3C+%281+-+z%29%2C+b+%3E+0%2C+n+%3E+0%2C+y+%3E+0%2C+y-b+%3E+0%2C+n+-+b+%3E+0%2C+0+%3C+z+%3C+0.5+solve+for+b+
+export function maximumRemovableLiquidity(pool: { [outcome: string]: number }) {
+  if (FRACTION_OF_POOL_MIN > 1 || FRACTION_OF_POOL_MIN < 0)
+    throw new Error('Math invariant violated')
+
+  const z = FRACTION_OF_POOL_MIN
+  const { YES: y, NO: n } = pool
+  const ratioMax = (n * z + y * z - Math.min(y, n)) / (2 * z - 1)
```
Comment (mantikoros) [inline common/src/calculate-cpmm.ts:615]: is this right? if x=y=100, then this value is negative, right?

Comment (mantikoros) [reply common/src/calculate-cpmm.ts:615]: I still don't really understand why we need a ratio limit if we already have a limit on the number of shares in the liquidity pool.

**Review by mantikoros — CHANGES_REQUESTED** [review]
Mostly good. I think the ratioMax stuff is not right, but otherwise LGTM.

File: backend/api/src/remove-liquidity.ts (hunk 2 above)
Comment (sipec) [reply backend/api/src/remove-liquidity.ts:83]: contract.token is type 'CASH' | 'MANA'

File: common/src/calculate-cpmm.ts (hunk 3 above)
Comment (sipec) [reply common/src/calculate-cpmm.ts:594]: this function is used on the frontend and I want it to still return what the pool values would be even if it is below 100

File: common/src/calculate-cpmm.ts (hunk 4 above)
Comment (sipec) [reply common/src/calculate-cpmm.ts]: ok, after playing around with the maniswap calculator a bit more, I'm convinced that extreme ratios are fine so I removed the restriction

**Pull request merged**